spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/servlet/RequestSlf4jMdcAutoConfiguration.java

```diff
@@ -0,0 +1,125 @@
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.slf4j.MDC;
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication.Type;
+import org.springframework.boot.web.servlet.FilterRegistrationBean;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.core.Ordered;
+import org.springframework.util.StringUtils;
+
+/**
+ * {@link EnableAutoConfiguration Auto-configuration} for the slf4j {@link MDC}
+ * providing the {@link HttpServletRequest} source (as 'src'), URI (as 'uri')
+ * and user name (as 'uid'). If one does not exist for these or is empty, the
+ * corresponding MDC entry will not be created. This runs as early as possible
+ * in order to make these settings available to as much logging as possible.
+ * 
+ * @author Bruce Brouwer
+ * @since 2.0
+ */
+@Configuration
+@ConditionalOnWebApplication(type = Type.SERVLET)
+@ConditionalOnClass(MDC.class)
+public class RequestSlf4jMdcAutoConfiguration {
+
+	@Bean
+	public FilterRegistrationBean<RequestSlf4jMdcFilter> requestSlf4jMdcFilter() {
+		final RequestSlf4jMdcFilter filter = new RequestSlf4jMdcFilter();
+		final FilterRegistrationBean<RequestSlf4jMdcFilter> registration = new FilterRegistrationBean<>(filter);
+		registration.setDispatcherTypes(EnumSet.of(DispatcherType.REQUEST, DispatcherType.ERROR, DispatcherType.ASYNC));
+		registration.setOrder(Ordered.HIGHEST_PRECEDENCE);
+		return registration;
+	}
+
+	public static class RequestSlf4jMdcFilter implements Filter {
+		private static final Log logger = LogFactory.getLog(RequestSlf4jMdcFilter.class);
+
+		@Override
+		public void init(final FilterConfig filterConfig) throws ServletException {
+			// nothing to configure
+		}
+
+		@Override
+		public void doFilter(final ServletRequest request, final ServletResponse response, final FilterChain chain)
+				throws IOException, ServletException {
+			try (Closeable src = putMDC("src", RequestSlf4jMdcFilter::src, request)) {
+				try (Closeable uri = putMDC("uri", RequestSlf4jMdcFilter::uri, request)) {
+					try (Closeable uid = putMDC("uid", RequestSlf4jMdcFilter::uid, request)) {
+						chain.doFilter(request, response);
+					}
+				}
+			}
+		}
+
+		private static String src(HttpServletRequest request) {
+			final String xff = request.getHeader("X-Forwarded-For");
+			return StringUtils.hasLength(xff) ? xff : request.getRemoteHost();
```

Comment:
If X-Forwarded-For is set, should I also append the remote host? If you put your application behind a load balancer, the remote host is going the be the address of the load balancer, which is not terribly useful. This can easily be mitigated in the load balancer by stripping out X-Forwaded-For headers from clients it does not trust. 

But if your application is not behind a load balancer that strips out malicious X-Forwarded-For headers, an attacker might be able to cover their tracks by injecting a fake X-Forwarded-For header. 

One solution would be this:
```
return StringUtils.hasLength(xff) ? xff + ", " + request.getRemoteHost() : request.getRemoteHost();
```
Alternatively, we could just put both in the MDC: both the X-Forwarded-For header and the remote host. Then apps could decide which of the MDC fields they want to add to their logging simply by how they configure logback (for example).